packages/venia-ui/lib/components/Extension/extension.js

```diff
@@ -0,0 +1,37 @@
+import React, { Fragment } from 'react';
+
+const extensionMap = new Map();
+
+/**
+ * Gets any extensions registered for an id and returns the component.
+ * @param {string} props.id id of the portal within the map
+ */
+export const Portal = ({ id }) => {
+    const components = extensionMap.get(id);
+    return <Fragment>{components}</Fragment>;
+};
+
+/**
+ * This extension component renders children into the DOM after React mounts.
+ *
+ * @example
+ * <Extension targetId={'extension-point-1'}>
+ *   <MyExtensionComponent>
+ * </Extension>
+ *
+ * @param {String} [props.targetId=root] id of injection target for extension
+ * @param {ReactNode} props.children
+ */
+export const Extension = props => {
+    const { children, targetId = 'root' } = props;
+
+    // Add new extensions.
+    const existing = extensionMap.get(targetId) || [];
+    existing.push(children);
```

Comment:
How safe is this?

Reply:
Safe from a security standpoint?

I feel yes, it bears the same security risks as the previous implementation.

Reply:
What content are you referring to? Can you give me an example?

Reply:
The children of `Extension`, which could be dynamic.

```jsx
const Dynamic = props => {
  const [clicked, setClicked] = useState(false)
  const text = clicked ? "yep" : "nope"

  const handleClick = useCallback(() => {
      setClicked(true)
  }, [setClicked])

  return (
    <div>
        <button onClick={handleClick}>
            click me
        </button>
        <Extension targetId="extension-point-1">
            {text}
        </Extension>
    </div>
  )
}
```

Reply:
In this PR we would only register extensions at the top level, at least in the examples I had so far. App/Portal mounting iterates through the extensions registered and then renders them, so it is entirely possible that if a component itself, within App, registers an `Extension` the portal for it may have already rendered.

Still I think to your question, if you had some Extension that was the output of some changing thing, would it update? I think so, as you see with the toast that uses user context to display if a user signs in after the fact.